fix(cache): invalidate `codeCache` in most cases when imports change

agilgur5 opened the pull request:
## Summary

The `codeCache` should be invalidated when imports change if declarations are needed, because imports' types can change the resulting declaration. 
- Fixes #292

**EDIT**: also added `!isolatedModules` to the check, see below comments regarding, e.g. enums, that could cause compiled JS code to change as well when imports change

## Details

- previously, the `checkImports` flag was set to `true` for type-checking, but `false` for compilation
  - I _believe_ it is `false` because the compiled JS shouldn't change if an import changes
    - though I'm not sure if that was the original intent behind the code
  - problematically though, compilation results can include declarations, and those _can_ change if imports change
    - for instance, the types of an import can change the declaration that is output
  - so now, only set it to `false` for compilation if declarations are _not_ needed

See my root cause analysis in https://github.com/ezolenko/rollup-plugin-typescript2/issues/292#issuecomment-1152951087

ezolenko commented:
I don't remember why it is this way, but I think your change is correct.

Now that I look at it closer, maybe we should always check imports regardless -- there might be cases where typescript generates different code based on what was imported. Maybe inlining enums into literals?

agilgur5 commented:
> Maybe inlining enums into literals?

OH, that's actually the perfect edge case! I was trying to think of an edge-case where checking imports would be necessary for just compiled JS code (i.e. not declarations) and couldn't come up with one. But I did suspect that there were other edge-cases.

In the RCA I compared it to Babel:

> This seems to be because the JS code output shouldn't change if an import changes. That, I believe, is accurate. That's one of the reasons why Babel can just transpile TS code per-file (whereas type-checking requires knowledge of the whole "Program", i.e. the whole module graph).

And, of course, `enums` are one of those things that Babel _can't_ handle!

TS does have a check for this though, [`isolatedModules`](https://www.typescriptlang.org/tsconfig#isolatedModules).

So what I can do is `checkImports` if `!isolatedModules || declaration`. That's going to be the majority of cases, tbh, but it leaves in some potential for optimization.

Could make that a separate PR though, as checking `declaration` is still certainly needed

agilgur5 commented:
added another commit to handle `!isolatedModules` case as well. modified the PR summary and title to match this as well.

think this is good to go now!
